File: nanopay/src/net/nanopay/approval/ApprovalDAO.java
package net.nanopay.approval;

import foam.core.FObject;
import foam.core.X;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.mlang.sink.Sum;

import static foam.mlang.MLang.*;

/**
 * When a single request is approved, checks all other approved requests and calculates points.
 * If points are sufficient ( >= requiredPoints ), removes all unused requests from approvalRequestDAO and
 * Re-submits the object to its dao (daoKey), where a rule should be defined to determine further actions.
 */
public class ApprovalDAO
  extends ProxyDAO {

  public ApprovalDAO(X x,DAO delegate) {
    setX(x);
    setDelegate(delegate);
  }

  @Override
  public FObject put_(X x, FObject obj) {
    DAO requestDAO = ((DAO)x.get("approvalRequestDAO"));
    ApprovalRequest old = (ApprovalRequest) requestDAO.find(obj);
    ApprovalRequest request = (ApprovalRequest) getDelegate().put(obj);
    if ( old != null && old.getStatus() != request.getStatus() ) {
      DAO requests = ApprovalRequestUtil.getAllRequests(x, request.getObjId(), request.getClassification());
      // if points are sufficient to consider object approved
      if ( getCurrentPoints(requests) >= request.getRequiredPoints() ||
      getCurrentRejectedPoints(requests) >= request.getRequiredRejectedPoints() ) {
        //removes all the requests that were not approved to clean up approvalRequestDAO
        removeUnusedRequests(requests);
        
        //puts object to its original dao
        rePutObject(x, request);
      }
    }
    return request;
  }

  private void rePutObject(X x, ApprovalRequest request) {
    DAO dao = (DAO) x.get(request.getDaoKey());
    FObject found = dao.inX(x).find(request.getObjId()).fclone();
    dao.inX(x).put(found);
  }

  private void removeUnusedRequests(DAO dao) {
    dao.where(EQ(ApprovalRequest.STATUS, ApprovalStatus.REQUESTED)).removeAll();
  }

  private long getCurrentPoints(DAO dao) {
    return ((Double)
      ((Sum) dao
        .where(EQ(ApprovalRequest.STATUS, ApprovalStatus.APPROVED))
        .select(SUM(ApprovalRequest.POINTS))
      ).getValue()
    ).longValue();
  }

  private long getCurrentRejectedPoints(DAO dao) {
    return ((Double)
      ((Sum) dao
        .where(EQ(ApprovalRequest.STATUS, ApprovalStatus.REJECTED))
        .select(SUM(ApprovalRequest.POINTS))
      ).getValue()
    ).longValue();
  }
}

File: src/foam/nanos/approval/ApprovalDAO.js
/**
 * @license
 * Copyright 2020 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.nanos.approval',
  name: 'ApprovalDAO',
  extends: 'foam.dao.ProxyDAO',

  javaImports: [
    'foam.core.FObject',
    'foam.core.X',
    'foam.dao.DAO',
    'foam.mlang.sink.Sum',
    'foam.nanos.auth.Subject',
    'foam.nanos.auth.User',
    'foam.nanos.dao.Operation',
    'foam.util.Auth',
    'foam.util.SafetyUtil',

    'static foam.mlang.MLang.*'
  ],

  javaCode: `
    public ApprovalDAO(X x, DAO delegate) {
      setX(x);
      setDelegate(delegate);
    }
  `,

  methods: [
    {
      name: 'put_',
      javaCode: `
        ApprovalRequest nu = (ApprovalRequest) obj;
        ApprovalRequest old = (ApprovalRequest) getDelegate().inX(x).find(nu.getId());
        ApprovalRequest request = (ApprovalRequest) (getDelegate().inX(x).put(obj)).fclone();

        if ( old != null && old.getStatus() != request.getStatus()
          || old == null && request.getStatus() != ApprovalStatus.REQUESTED
        ) {
          DAO requests = ApprovalRequestUtil.getAllRequests(x, getDelegate().inX(x), request.getObjId(), request.getClassification());
          // if not a cancellation request and points are sufficient to consider object approved
          var currentPoints = getCurrentPoints(requests);
          var currentRejectedPoints = getCurrentRejectedPoints(requests);
          if (
            request.getStatus() == ApprovalStatus.CANCELLED ||
            currentPoints >= request.getRequiredPoints() ||
            currentRejectedPoints >= request.getRequiredRejectedPoints()
          ) {

            //removes all the requests that were not approved to clean up approvalRequestDAO
            removeUnusedRequests(requests);
            if (
              request.getStatus() == ApprovalStatus.APPROVED ||
              (
                request.getStatus() == ApprovalStatus.REJECTED && ((ApprovalRequest) request).getOperation() == Operation.CREATE
              )
            ){
              //puts object to its original dao
              try {
                rePutObject(x, request);

                // unassign request once resolved
                request.setAssignedTo(0);
                getDelegate().put(request);

              } catch ( Exception e ) {
                request.setStatus(ApprovalStatus.REQUESTED);
                getDelegate().put(request);
                throw new RuntimeException(e);
              }
            } else {
              // since no more needs to be done with the request from this point onwards
              request.setIsFulfilled(true);
              request.setAssignedTo(0);
              getDelegate().put(request);
            }
          }
        }
        return request;
      `
    },
    {
      name: 'rePutObject',
      visibility: 'protected',
      type: 'Void',
      args: [
        { type: 'Context', name: 'x' },
        { type: 'ApprovalRequest', name: 'request' }
      ],
      javaCode: `
        String daoKey = request.getServerDaoKey() != null && ! SafetyUtil.isEmpty(request.getServerDaoKey()) ? request.getServerDaoKey() : request.getDaoKey();
        DAO dao = (DAO) x.get(daoKey);
        FObject obj = dao.find(request.getObjId()).fclone();

        DAO userDAO = (DAO) x.get("localUserDAO");
        User initiatingAgent = null;
        if ( ((ApprovalRequest) request).getCreatedForAgent() != 0 ) {
          initiatingAgent = (User) userDAO.find(((ApprovalRequest) request).getCreatedForAgent());
        }

        User initiatingUser = (User) userDAO.find(((ApprovalRequest) request).getCreatedBy());
        if ( ((ApprovalRequest) request).getCreatedFor() != 0 ) {
          initiatingUser = (User) userDAO.find(((ApprovalRequest) request).getCreatedFor());
        }

        if ( initiatingAgent == null ) {
          initiatingAgent = initiatingUser;
        }

        X initiatingUserX = Auth.sudo(x, initiatingUser, initiatingAgent);

        User realUser = ((Subject) x.get("subject")).getRealUser();
        User user = ((Subject) x.get("subject")).getUser();
        Subject initiatingUserSubject = (Subject) initiatingUserX.get("subject");
        initiatingUserSubject.getUserPath().add(realUser);
        if ( realUser != user ) {
          initiatingUserSubject.getUserPath().add(user);
        }

        initiatingUserX = initiatingUserX.put(ApprovalRequest.class, request);

        if ( ((ApprovalRequest) request).getOperation() == Operation.REMOVE ) {
          dao.inX(initiatingUserX).remove(obj);
        } else {
          dao.inX(initiatingUserX).put(obj);
        }
      `
    },
    {
      name: 'removeUnusedRequests',
      visibility: 'protected',
      type: 'Void',
      args: [
        { type: 'DAO', name: 'dao' }
      ],
      javaCode: `
        dao.where(EQ(ApprovalRequest.STATUS, ApprovalStatus.REQUESTED)).removeAll();
      `
    },
    {
      name: 'getCurrentPoints',
      visibility: 'protected',
      type: 'Long',
      args: [
        {
          type: 'DAO', name: 'dao' }
      ],
      javaCode: `
        return ((Double)
          ((Sum) dao
            .where(EQ(ApprovalRequest.STATUS, ApprovalStatus.APPROVED))
            .select(SUM(ApprovalRequest.POINTS))
          ).getValue()
        ).longValue();
      `
    },
    {
      name: 'getCurrentRejectedPoints',
      visibility: 'protected',
      type: 'Long',
      args: [
        { type: 'DAO', name: 'dao' }
      ],
      javaCode: `
        return ((Double)
          ((Sum) dao
            .where(EQ(ApprovalRequest.STATUS, ApprovalStatus.REJECTED))
            .select(SUM(ApprovalRequest.POINTS))
          ).getValue()
        ).longValue();
      `
    }
  ]
});
